feat: Add support for showing and hiding parts of the progress page

xitij2000 opened the pull request:
This PR adds support for hiding or showing components on the progress page based on configuration stored in `other_course_settings`. 

These settings can be toggled with: https://github.com/open-craft/frontend-app-authoring/pull/71

Private-ref: [BB-9200](https://tasks.opencraft.com/browse/BB-9200)

RafayGhafoor committed:
fix: make progress graph respect course settings (#1194)

(cherry picked from commit 391ea08b20dfa425e1a2a6622ae742b22731cd91)

xitij2000 committed:
feat: Add support for showing and hiding parts of the progress page

Agrendalath reviewed (commented):
@xitij2000, please see the following recording. We cannot have this kind of delay when hiding the elements on the page:
[Screencast_20241008_182416.webm](https://github.com/user-attachments/assets/b307213e-1004-4c03-885e-fdfc7ef79968)


I just noticed that [there is already a model](https://github.com/openedx/edx-platform/blob/d9a91d4c38d8c7d8fcc41a2bcc042470b0b9920a/lms/djangoapps/course_home_api/models.py#L11) with a per-course Progress page configuration that's [used](https://github.com/openedx/edx-platform/blob/a389a9ff10fd06bd053107fc6a4f19c7d24f7233/lms/djangoapps/course_home_api/progress/views.py#L181-L182) by the Course Home API. If we would need to add artificial delays to rendering the interface until the Course Blocks API gives us the response, maybe extending this configuration would actually be better?

Also, a note about the Pages & Resources - the options should either:
1. Default to True.
2. Be "Hide the component" instead of "Show the component".

Otherwise, visiting the "Configure progress" modal and pressing "Save" changes the Progress page configuration, which is confusing.

xitij2000 commented:
@Agrendalath Sure, I think it makes sense to hide by default. I did intend in the code to default an undefined value to true, but I guess that didn't work? The logic should be to hide till the value is known, but have an undefined value equate to true so that if this isn't configured, then the component is shown by default. 

The issue isn't with extending that configuration, it's with extending it in a way that is upstream-friendly. Making a change to an existing model is asking for trouble during upgrades when we need to cherrypick migrations etc. 

If upstream is okay with having this feature, then we can move this to the man course structure and return it with the existing API.

Agrendalath commented:
@xitij2000,
> Sure, I think it makes sense to hide by default. I did intend in the code to default an undefined value to true, but I guess that didn't work? The logic should be to hide till the value is known, but have an undefined value equate to true so that if this isn't configured, then the component is shown by default.

What are the next steps here?

> The issue isn't with extending that configuration, it's with extending it in a way that is upstream-friendly. Making a change to an existing model is asking for trouble during upgrades when we need to cherrypick migrations etc.

If needed, we can backport this change only for this client and handle it during the upgrade. It won't be upstream-unfriendly this way.

xitij2000 committed:
fixup! feat: Add support for showing and hiding parts of the progress page

xitij2000 commented:
@Agrendalath 
> What are the next steps here?

I've hopefully already fixed this issue in the latest commit. 

> > The issue isn't with extending that configuration, it's with extending it in a way that is upstream-friendly. Making a change to an existing model is asking for trouble during upgrades when we need to cherrypick migrations etc.
> 
> If needed, we can backport this change only for this client and handle it during the upgrade. It won't be upstream-unfriendly this way.

I think I miswrote. It maybe is upstream friendly, but it isn't upgrade-friendly. Maintaining diverging migration histories is not a good idea, I feel. 

Also, that config is derived from `StackedConfigurationModel` which will make it unnecessarily complex. The `StackedConfigurationModel` allows overriding values on a global, site, org or course, or course run level, so you could have a page disabled globally, enabled for the site, disabled for the org, then enabled again for the course, and disabled again for the course run, and it will follow that hierarchy and give you the closest matching status. That level of complexity isn't needed here and won't be supported by the GUI. Stacked configs are generally only applied via the django admin and not interacted with by APIs. So I don't see this ever being upstream-friendly, which means it's likely to be a permanent drift, or an expensive rollback and change later. 

It would be better to add a separate model for this if needed, but in that case, it won't be part of course reruns or import-export, which might be OK.

Agrendalath reviewed (changes requested):
@xitij2000, the problem with disappearing elements is resolved, but there are some more issues:
1. There is no way to hide the "Course completion" section.
2. When the grades are hidden, an empty `.grades.raised-card` node is still visible.
3. The `CertificateStatus` option is not implemented in [this PR](https://github.com/open-craft/frontend-app-authoring/pull/71). The `[data-testid=certificate-status-component]` node is always visible.
4. The second part of [this comment](https://github.com/open-craft/frontend-app-learning/pull/36#pullrequestreview-2355013586) is still not addressed.

xitij2000 committed:
fixup! fixup! feat: Add support for showing and hiding parts of the progress page

xitij2000 commented:
> @xitij2000, the problem with disappearing elements is resolved, but there are some more issues:
> 
>     1. There is no way to hide the "Course completion" section.

This is a feature that already existed for the progress page, it's called "Enable progress graph"
> 
>     2. When the grades are hidden, an empty `.grades.raised-card` node is still visible.

I've changed it, so the entire component is hidden now. 

> 
>     3. The `CertificateStatus` option is not implemented in [this PR](https://github.com/open-craft/frontend-app-authoring/pull/71). The `[data-testid=certificate-status-component]` node is always visible.
> 

I seems that got dropped during some rebase, I've added it back. 

>     4. The second part of [this comment](https://github.com/open-craft/frontend-app-learning/pull/36#pullrequestreview-2355013586) is still not addressed.

Fixed as well.

Agrendalath reviewed (changes requested):
@xitij2000, after the most recent change, the "Show grades" toggle does not do anything. The "Show Grade Breakdown" toggle hides the entire `.grades.rounded` node.

> This is a feature that already existed for the progress page, it's called "Enable progress graph"

I see. I cherry-picked https://github.com/openedx/frontend-app-learning/pull/1194 into your branch,so that we can use it in Palm. I also opened https://github.com/open-craft/edx-platform/pull/698 to backport the backend changes.

xitij2000 commented:
@Agrendalath Not sure why the show grades toggle is doing nothing. I think I'll fix up my palm devstack since right now I'm not able to test this myself.

xitij2000 commented:
> @Agrendalath Not sure why the show grades toggle is doing nothing. I think I'll fix up my palm devstack since right now I'm not able to test this myself.

I fixed up my palm devstack and the show grades toggle is working fine for me? It shows and hides the "Grades" widget. Could check again? Are you using tutor or is this a different setup?

Agrendalath commented:
@xitij2000, ah, sorry. I forgot we're hiding this section with the branding package. Everything's good here. Please add the description to the PR and link the related PRs and tickets so that we can easily navigate between all related work.

xitij2000 commented:
@Agrendalath  Done!

xitij2000 committed:
fixup! fix: make progress graph respect course settings (#1194)

Agrendalath commented:
This is no longer needed, as we don't use Palm anymore. Closing.
